services/horizon/internal/actions: Query struct with gorilla/schema and govalidators.

abuiles opened the pull request:
Automate query params reading and validation. 

## Usage

Define your struct, specifying the url schema tags and validations as defined in [schema](https://www.gorillatoolkit.org/pkg/schema) and [govalidator](https://github.com/asaskevich/govalidator).

Let's suppose you are working on an end-point that requires a  Stellar account, like `/accounts/{account_id}/offers`, the following struct will pull the required data from the query params and validate that it is a valid address.

```go
type AccountOffersQuery struct {
	AccountID string `schema:"account_id" valid:"accountID"`
}
```

Use it in the context of an account like:

```go
func (handler GetAccountOffersHandler) GetResourcePage(r *http.Request) ([]hal.Pageable, error) {
	ctx := r.Context()
	qp := OffersQuery{}
	qp := AccountOffersQuery{}
	err := GetParams(&qp, r)
	if err != nil {
		return nil, err
	}

	accountID := xdr.AccountId{}
	if err := accountID.SetAddress(qp.Address); err != nil {
		panic(err) // this shouldn't err since GetParams runs the validator 
	}
         //  .... do something with your account

}
```

### Shared query params

#### `SellingBuyingAssetQueryParams`

This is a struct meant to be embedded and used with actions which have end-point that required a selling and buying asset, like `/offers`. 

Given the following struct:

```go
// OffersQuery query struct for offers end-point
type OffersQuery struct {
	SellingBuyingAssetQueryParams `valid:"-"`
	Seller string `schema:"seller" valid:"accountID"`
}
```

See how you can get the selling and buying asset by calling `struct.Selling()` and `struct.Buying()` -- validation will be done in `GetParams`

```go
// GetResourcePage returns a page of offers.
func (handler GetOffersHandler) GetResourcePage(r *http.Request) ([]hal.Pageable, error) {
	ctx := r.Context()
	qp := OffersQuery{}
	err := GetParams(&qp, r)

	if err != nil {
		return nil, err
	}

	query := history.OffersQuery{
		SellerID:  qp.Seller,
		Selling:   qp.Selling(),
		Buying:    qp.Buying(),
	}

	offers, err := getOffersPage(ctx, handler.HistoryQ, query)
	if err != nil {
		return nil, err
	}

	return offers, nil
}
```

## Validators

This PR includes the following query params validators:

#### `accountID`

`accountID` validates if a query param is a valid stellar address.

### `assetType`

`accountID` validates if a query param is a valid asset type, that is: `native`, `credit_alphanum4` or `credit_alphanum12`.

bartekn commented:
I haven't checked the code yet but regarding your comment I think we can automate a lot of things so the code will be smaller than in your snippets:
* Automating validation removes a need for checking for errors when getting custom type variables. For example, if you add a validation tag for seller:
   ```go
   Seller string `schema:"seller" validate:"account_id"`
   ```
   the `SellerAccountID()` method will never return error because it's a valid account ID. Same thing applies to assets and any other custom type.
* For complex types you can have a custom `Validate()` method on query struct that will contain the logic required to validate the query properly.

So ultimately, the offers example can look like this:
```go
 type OffersQuery struct { 
 	QueryParams 
 	Seller              string `schema:"seller" validate:"account_id"` 
 	SellingAssetType    string `schema:"selling_asset_type"` 
 	SellingAsssetIssuer string `schema:"selling_asset_issuer"` 
 	SellingAsssetCode   string `schema:"selling_asset_code"` 
 	BuyingAssetType     string `schema:"buying_asset_type"` 
 	BuyingAsssetIssuer  string `schema:"buying_asset_issuer"` 
 	BuyingAsssetCode    string `schema:"buying_asset_code"` 
 } 

func (q OffersQuery) Validate() error {
    err := validator.Validate(q) // Global validation for known types, ex. account_id
    if err != nil {
        return err
    }

    // Custom code here: ex. Selling and Buying asset but we can also
    // develop more sophisticated code that is able to automatically
    // validate a group of fields.
}

// Same for Buying. Please note that there are several places
// where we use `selling_*` and `buying_` so we can extract it
// to remove code duplication.
func (q OffersQuery) Selling() (*xdr.Asset) {
    // If selling not required and fields empty return `nil`.
    selling, err := BuildAsset(q.SellingAssetType, q.SellingAsssetIssuer, q.SellingAsssetCode)
    if err != nil {
        panic(err) // We validated it before so it should never panic here.
    }
    return selling
}

// With OfferQuery like above the action code will look like this:

func (handler GetOffersHandler) GetResourcePage(r *http.Request) ([]hal.Pageable, error) { 
 	ctx := r.Context() 
 	qp := OffersQuery{} 
 	err := GetParamsValidate(&qp, r) // gets params and runs `Validate()` automatically
 	if err != nil {
 		// err is Problem if query is not valid
 		return nil, err 
 	} 
  
 	query := history.OffersQuery{ 
 		PageQuery: qp.PageQuery(), 
 		SellerID:  qp.Seller(), // or qp.Seller if you need a string
 		Selling:   qp.Selling(), // Will return `nil` if not required and fields empty
 		Buying:    qp.Buying(), 
 	}

 	// Run query and return `[]Offers`
}

```

abuiles commented:
@bartekn I see your point, I'll complete this once I'm back and we can compare both solutions.  

What I noticed as I was working on this is that we were actually requiring more code to deal with query params, but maybe that's also because I took as example a route with a bunch of edge cases.  For other scenarios I agree this reduces the amount of boiler plate, the following for example:

```go
func (handler GetAccountOffersHandler) parseOffersQuery(r *http.Request) (history.OffersQuery, error) {
	pq, err := GetPageQuery(r)
	if err != nil {
		return history.OffersQuery{}, err
	}

	seller, err := GetString(r, "account_id")
	if err != nil {
		return history.OffersQuery{}, err
	}

	query := history.OffersQuery{
		PageQuery: pq,
		SellerID:  seller,
	}

	return query, nil
}

// GetResourcePage returns a page of offers for a given account.
func (handler GetAccountOffersHandler) GetResourcePage(r *http.Request) ([]hal.Pageable, error) {
	ctx := r.Context()
	query, err := handler.parseOffersQuery(r)
	if err != nil {
		return nil, err
	}

	offers, err := getOffersPage(ctx, handler.HistoryQ, query)
	if err != nil {
		return nil, err
	}

	return offers, nil
}
```

Will become 

```go
type AccountOffersQuery struct {
	QueryParams
	Seller              string `schema:"seller" validate:"account_id"`
}

// GetResourcePage returns a page of offers for a given account.
func (handler GetAccountOffersHandler) GetResourcePage(r *http.Request) ([]hal.Pageable, error) {
	ctx := r.Context()
	pq := AccountOffersQuery{}
	err := GetParamsValidate(pq, r)
	if err != nil {
		return nil, err
	}
	query := history.OffersQuery{
		PageQuery: pq,
		SellerID:  pq.Seller,
	}

	offers, err := getOffersPage(ctx, handler.HistoryQ, query)
	if err != nil {
		return nil, err
	}

	return offers, nil
}
```
We could even cut the request earlier through a middleware or something if query params are not valid.

abuiles commented:
@bartekn @tamirms can you have a second look at this? I implemented the final version using gorilla schema and govalidator. It also adds a custom validator for the scenarios where we have buying and selling asset. 

I still need to add more validations for the page query parameters but I think with this you can see the pattern now. 

Please take a look and let me know, I've left some comments on the code and TODOs for things that still need to be addressed. I'll tackle those ones after your review.

abuiles commented on services/horizon/internal/actions/helpers.go:
```diff
@@ -701,3 +760,62 @@ func FullURL(ctx context.Context) *url.URL {
+		}
+	}
+
+	if err := decoder.Decode(dst, query); err != nil {
```
we take the value here with gorilla schema.

abuiles commented on services/horizon/internal/actions/helpers.go:
```diff
@@ -701,3 +760,62 @@ func FullURL(ctx context.Context) *url.URL {
+		return err
+	}
+
+	if _, err := govalidator.ValidateStruct(dst); err != nil {
```
once the values are out, we run the validations.

abuiles commented on services/horizon/internal/actions/helpers.go:
```diff
@@ -701,3 +760,62 @@ func FullURL(ctx context.Context) *url.URL {
+	if _, err := govalidator.ValidateStruct(dst); err != nil {
+		fields := govalidator.ErrorsByField(err)
+
+		for field, message := range fields {
```
the field value, would be as defined in the struct, but we want to actually report the error as defined in the query params.  so instead of `BuyingAssetCode` it will use the value defined in the schema which was the one passed in the request `buying_asset_code`

abuiles commented on services/horizon/internal/actions/helpers_test.go:
```diff
@@ -569,6 +570,130 @@ func TestFullURL(t *testing.T) {
+	tt.Assert.Equal("desc", qp.Order)
+	tt.Assert.Equal(2, qp.Limit)
+	tt.Assert.Equal(account, qp.Account)
+	tt.Assert.Equal(usd.String(), qp.Selling().String()) // TODO: how can I compare usd with *Asset?
```
what's the best approach to compare asset with *asset ?

tamirms replied:
`usd.Equals(qp.Selling())`

abuiles commented on services/horizon/internal/actions/offer.go:
```diff
@@ -3,15 +3,117 @@ package actions
 
+// QueryParams query struct for pagination params
+// TODO: move a shared package - maybe even db2
+type PageQueryParams struct {
```
we can move this out to a different file -- maybe helpers?

abuiles commented on services/horizon/internal/actions/offer.go:
```diff
@@ -3,15 +3,117 @@ package actions
+
+// SellingBuyingAssetQueryParams query struct for end-points requiring a selling
+// and buying asset
+type SellingBuyingAssetQueryParams struct {
```
This will be reusable by other end-point with similar structure, what would be a good file to put this shared query params? helpers? Also should we keep them in this package?

abuiles commented on services/horizon/internal/actions/offer.go:
```diff
@@ -3,15 +3,117 @@ package actions
+		panic(err)
+	}
+
+	// TODO: move this validation to validator
```
I'll add validators on the query params next so that we don't need this here

abuiles commented on services/horizon/internal/actions/validators.go:
```diff
@@ -0,0 +1,94 @@
+func InitValidators() {
+	govalidator.TagMap["accountID"] = govalidator.Validator(isAccountID)
+	govalidator.TagMap["assetType"] = govalidator.Validator(isAssetType)
+	govalidator.CustomTypeTagMap.Set("sellingCode", isValidSellingCode)
```
this and the next validators are valid for `SellingBuyingAssetQueryParams` struct

tamirms commented on services/horizon/internal/actions/helpers.go:
```diff
@@ -442,6 +454,53 @@ func getAssetType(r *http.Request, name string) (xdr.AssetType, error) {
 
+// BuildAsset creates a new asset from a given asset_type, asset_code, and asset_issuer.
+func BuildAsset(assetType, issuer, code string) (xdr.Asset, error) {
+	var value interface{}
```
I think this could be simplified as:

```golang
var asset xdr.Asset
switch t {
	case xdr.AssetTypeAssetTypeNative:
		if err := asset.SetNative(); err != nil {
			return xdr.Asset{}, err
		}
	default:
		issuerAccountID := xdr.AccountId{}
		if err := issuerAccountID.SetAddress(issuer); err != nil {
			return xdr.Asset{}, err
		}

		if err := asset.SetCredit(code, issuerAccountID); err != nil {
			return xdr.Asset{}, err
		}
}
return asset, nil
```

abuiles replied:
@tamirms good call! 

I might be wrong, but I think the reason this is built like it was is because there could  be a scenario where you have a `credit_alphanum12` but something like `FOO` in the code - or do we force length to be between 5 and 12 when `credit_alphanum12` is specified? 

If the asset I just described is valid, then `asset.SetCredit(code, issuerAccountID)` will produce the wrong result since it will pick up `credit_alphanum4`, which will generate the incorrect base64 representation when calling `xdr.MarshalBase64()` 

cc @bartekn

abuiles replied:
@bartekn sounds good! does that mean we could simplify query params when receiving assets? Right now we require `asset_type`, `asset_code` and `asset_issuer` in different end-points -- is there a particular reason for requiring the type? Or can we remove that and take only `asset_code` and `asset_issuer`?

bartekn replied:
> is there a particular reason for requiring the type?

This is a good question. `asset_type` param is currently redundant because you can derive it from the code length. The main reason for passing `asset_type` is because if there's a new asset type with code and issuer it will be used to distinguish them. On the other hand, looks like existing asset types serve most of the use cases and there are no other asset types planned AFAIK. Also, we introduced `CODE:ISSUER` format in `/paths` (that is not consistent with the one here).

> Or can we remove that and take only asset_code and asset_issuer?

I think we shouldn't do it in this PR for sure because it would be a breaking change. Let's discuss this in Horizon meeting.

abuiles replied:
> I think we shouldn't do it in this PR for sure because it would be a breaking change. Let's discuss this in Horizon meeting.

Since this is a new end-point, we can use the new format without introducing  a breaking change.

abuiles replied:
👍 applied @tamirms's suggestion and moved to xdr.asset

tamirms commented on services/horizon/internal/actions/helpers.go:
```diff
@@ -701,3 +760,62 @@ func FullURL(ctx context.Context) *url.URL {
+	// contain `account_id` and URL Query params will contain `foo`.
+	if rctx := chi.RouteContext(r.Context()); rctx != nil {
+		for _, key := range rctx.URLParams.Keys {
+			query.Set(key, rctx.URLParam(key))
```
should we return an error if a route param collides with an existing query param?

abuiles replied:
sounds like a good idea, we should probably panic in that scenario

abuiles replied:
@tamirms I was thinking about this and if there is a collision, I think we should give priority to chi's query params and ignore the passed-in value since it isn't probably one of the defined (expected) query params.

If we define a route like `/accounts/{account_id}/offers`, it's likely we won't have an `account_id` as one of the possible query params, if the request contains it, then we just ignore it.

tamirms commented on services/horizon/internal/actions/helpers_test.go:
```diff
@@ -569,6 +570,130 @@ func TestFullURL(t *testing.T) {
+}
+
+func (pv ParamsValidator) Validate() error {
+	return problem.MakeInvalidFieldProblem(
```
should this always return an error? it seems like it's missing some validation logic

tamirms replied:
nevermind, I see that it's only used for tests

tamirms commented on services/horizon/internal/actions/validators_test.go:
```diff
@@ -0,0 +1,210 @@
+)
+
+func TestMain(m *testing.M) {
+	InitValidators()
```
I would prefer if `InitValidators()` is called at the beginning of each test instead of using `TestMain` because `TestMain` is applied on all tests in a package even if the tests are defined in a different file

tamirms commented on services/horizon/internal/actions/query_params.go:
```diff
@@ -0,0 +1,124 @@
+
+	if err != nil {
+		// should have been valid
+		panic(err)
```
I'd prefer if we could avoid panicking. One alternative is to include the parsed limit in the struct and have a `Validate()` function populate it:

```golang
type PageQueryParams struct {
	Cursor string `schema:"cursor" valid:"cursor~the value should not be a negative number"`
	Order  string `schema:"order" valid:"in(asc|desc)~valid values are asc or desc"`
	RawLimit  string `schema:"limit" valid:"int,range(1|200)~value should be between 1 and 200"`
	Limit uint64
}

func (params *PageQueryParams) Validate() error {
	...
	params.Limit, err = strconv.ParseInt(params.RawLimit, 10, 64)
	...
}
```

abuiles replied:
@tamirms we should "never" reach this state, since validate will avoid getting to this if the param is invalid.

abuiles replied:
@tamirms we have an `int` validation on limit, so by the time the code is reached we know that ParseInt will succeed. That' why we panic if there is an `err`. 

That said though, I think there is nothing avoiding people to do something like:

```
pq = PageQueryParams{...}
pq.getLimit() // <-- if this is called without running the validation first, then we can't guarantee it is an int 
```

tamirms replied:
I understand that an error shouldn't arise because of the validation tags on the struct. however, in case there's a mismatch between the validator function and the parse function it would be nice if horizon handled it rather than just crashing

abuiles replied:
@tamirms makes sense,  we could do that but then we'll loose the ergonomics of query structs, we'll be back to doing err handing, which we were leaving to govalidator.

@bartekn what are your thoughts on this?

abuiles replied:
@bartekn makes sense! I think I went with string to be able to identify between 0 and not passed. I think we can assume that if limit is 0, then it was not passed and use the default.

abuiles replied:
@bartekn @tamirms please check this commit which addresses this issue. I didn't end up doing this ` I think we can assume that if limit is 0` , if you pass zero, a negative value or a number higher than two hundred then it will be invalid. 

https://github.com/stellar/go/pull/1755/commits/03d075cfb7772c3b5c05e9ba7e2aae6ffaeb65cb

bartekn replied:
It's ok. I think we don't need to be so strict about about non-set values. Something like:
```go
Limit int `schema:"limit" valid:"int,range(0|200)~value should be between 1 and 200"`
```
would work too. In all options (string, *int or int with `0` allowed) we need a code that translates zero value to default anyway.

abuiles replied:
@bartekn  the solution in https://github.com/stellar/go/commit/03d075cfb7772c3b5c05e9ba7e2aae6ffaeb65cb requires a little bit more of extra work but I'm actually happy with it and avoids a possible panic scenario. 

I was thinking about converting 0 to default value, but I think this could end up confusing/surprising users. If someone is sending `limit=0` we should `400` instead of rewriting the value and ack'ing the request.  It could be confusing for people to ask for 0 values and get back 10.

abuiles commented on services/horizon/internal/actions/query_params.go:
```diff
@@ -0,0 +1,124 @@
+// SellingBuyingAssetQueryParams query struct for end-points requiring a selling
+// and buying asset
+type SellingBuyingAssetQueryParams struct {
+	SellingAssetType   string `schema:"selling_asset_type" valid:"assetType~valid types are native credit_alphanum4 or credit_alphanum12"` // TODO using a comma doesn't work in custom message, figure how to make it work
```
I'll add a map to validators.go so we can format the message correctly. The reason why commas don't work is because that's the syntax to add join multiple validators `validation:"int,range(1|10),isEven"`

bartekn commented on services/horizon/internal/actions/query_params.go:
```diff
@@ -0,0 +1,124 @@
+	BuyingAssetType    string `schema:"buying_asset_type" valid:"assetType"`
+	BuyingAssetIssuer  string `schema:"buying_asset_issuer" valid:"accountID"`
+	BuyingAssetCode    string `schema:"buying_asset_code" valid:"buyingCode"`
+}
```
I think we should change validation of this struct a little.  `SellingAssetCode` and `BuyingAssetCode` should use `valid:"assetCode"` and then we need `SellingBuyingAssetQueryParams.Validate()` method that will check if `SellingAssetType` and `SellingAssetCode` are compatible. We also need to remove `sellingCode` and `buyingCode` validators because they depend on other fields of the struct.

abuiles replied:
solved via 4da363e80d813a517bcc29035fef54b75567114c

bartekn commented on services/horizon/internal/actions/validators.go:
```diff
@@ -0,0 +1,126 @@
+	govalidator.SetFieldsRequiredByDefault(false)
+}
+
+func validatorErrorFor(msg string) string {
```
It sucks that we can't return custom errors (instead of `bool`) in validation functions in `govalidator`... I don't like string comparisons like this because it's easy to miss during refactoring or package upgrade. Can we instead `switch` of validator's name (ex. `assetType`, `assetCode`)?

abuiles replied:
agree with you, I look briefly and there was no easy way to get the validator, I'll have a second look though. It might be possible.

abuiles replied:
I didn't think hard enough  ;) -- solved via https://github.com/stellar/go/pull/1755/commits/2615c07338cbb071409e93224d3a5b7b95cd49f3

bartekn commented on services/horizon/internal/actions/validators.go:
```diff
@@ -0,0 +1,126 @@
+}
+
+func isAssetType(str string) bool {
+	if _, err := assets.Parse(str); err != nil {
```
👍 but we should really change `assets.Parse` to `assets.ParseType`.

abuiles commented on services/horizon/internal/actions/query_params.go:
```diff
@@ -0,0 +1,124 @@
+
+	if lastEventID := r.Header.Get("Last-Event-ID"); lastEventID != "" {
+		cursor = lastEventID
+	}
```
@bartekn good point!

abuiles replied:
@bartekn after thinking more about this, I decided to take a different approach,  using query struct for all query params but the pagination related ones. 

I think with this we still achieve our biggest goal of simplifying query params reading/validation, but for page query there are a couple of scenarios that make it a bit awkward to generalize, specifically:

  - We rely on data from the request to override cursor
  - There might be scenarios where we want to do cursor validation or not, which is done at the action level and then we need to send that data down somehow to `db2.NewPageQuery`

See 3d7568d643c0546bb3a4abed4449f50b3066dc88

abuiles replied:
commit change to 64a87f2f

bartekn commented:
> I was thinking about converting 0 to default value, but I think this could end up confusing/surprising users. If someone is sending limit=0 we should 400 instead of rewriting the value and ack'ing the request. It could be confusing for people to ask for 0 values and get back 10.

You're right. (Somehow I can't reply in the thread...)

abuiles commented on services/horizon/internal/actions/query_params.go:
```diff
@@ -0,0 +1,141 @@
+
+// PageQueryParams query struct for pagination params
+type PageQueryParams struct {
+	Cursor string `schema:"cursor" valid:"cursor~the value should not be a negative number"`
```
@bartekn I forgot `cursor` can  take multiple shapes and that we can opt-out of cursor validation. This generalization doesn't work here, we'll have to let the consumer (an action's `GetResourcePage`) decide if they want to run cursor validation or not when they get the PageQuery -- something like

```go
	pq, err := qp.PageQuery(r, DisableCursorValidation)
	if err != nil {
		return nil, err
	}

	query := history.OffersQuery{
		PageQuery: pq,
		SellerID:  qp.Seller,
		Selling:   qp.Selling(),
		Buying:    qp.Buying(),
	}
```

abuiles commented:
@bartekn @tamirms I addressed all your comments, please check again. 

I'll squash and rebase to resolve the conflicts once the PR is approved. I want to use some of the existing commits to help with some comments and discussions happening in the PR>

bartekn commented on services/horizon/internal/actions/validators.go:
```diff
@@ -0,0 +1,92 @@
+	var validLen int
+	switch t {
+	case xdr.AssetTypeAssetTypeNative:
+		return nil
```
We should actually check if code and issuer are empty in this case. Otherwise, strange queries like `type=native&code=USD` will be accepted.

bartekn commented on services/horizon/internal/actions/query_params.go:
```diff
@@ -0,0 +1,88 @@
+
+// Validate runs validations on asset_code for buying and selling
+func (q SellingBuyingAssetQueryParams) Validate() error {
+	if len(q.SellingAssetType) > 0 {
```
We should probably also return `400` error for queries like `selling_asset_code=USD&selling_asset_issuer=[valid_issuer]` (user forgets to pass type). It will succeed everywhere (because `len(q.SellingAssetType) = 0`) but filtering won't work.

abuiles replied:
makes sense, I added a bunch of validations for the following matrix

|Type|Code|Issuer|
|----|----|------|
|X   | X  |  X   |
|X   | X  |      |
|X   |    |      |
|    | X  |      |
|    | X  |  X   |
|    |    |  X   |

bartekn commented on services/horizon/internal/actions/query_params_test.go:
```diff
@@ -0,0 +1,122 @@
+				"selling_asset_type":   "credit_alphanum4",
+				"selling_asset_code":   "EUR",
+				"selling_asset_issuer": "GBRPYHIL2CI3FNQ4BXLFMNDLFJUNPU2HY3ZMFSHONUCEOASW7QC7OX2H",
+			},
```
Queries like `selling_asset_code=USD&selling_asset_issuer=[valid_issuer]` should fail too.

bartekn commented on services/horizon/internal/actions/validators.go:
```diff
@@ -0,0 +1,92 @@
+	// govalidator handles embedded structs as fields and tries to find a
+	// validator for them, failing if it doesn't find one. This relaxes this
+	// setup a bit, so it doesn't try to validate on the embedded struct.
+	govalidator.SetFieldsRequiredByDefault(false)
```
Can you check if this is in any conflict with https://github.com/stellar/go/blob/613572f4783969f8193f582331dd28bd6e52d1ac/support/config/main.go#L65?  I think we import this package in Horizon.

abuiles replied:
good call! I actually remove this and let's use the default which is true. 

We just need to mark explicitly whether a field is required or optional in the query params. I like that approach better,  since it's more explicit about the expectations.

bartekn commented on services/horizon/internal/actions/validators.go:
```diff
@@ -0,0 +1,92 @@
+		validLen = len(xdr.AssetAlphaNum12{}.AssetCode)
+	}
+
+	if len(code) > validLen {
```
Looks like empty code will pass this test?

abuiles replied:
Fixed

tamirms commented on xdr/asset.go:
```diff
@@ -32,6 +32,43 @@ func MustNewCreditAsset(code string, issuer string) Asset {
+
+	var asset Asset
+	switch t {
+	case AssetTypeAssetTypeNative:
```
should we check that `issuer` and `code` are empty in this case?

abuiles replied:
@tamirms good call, Bartek mentioned the same above, I just added a regression test for this scenario :)

tamirms commented on xdr/account_id.go:
```diff
@@ -84,6 +84,16 @@ func MustAddress(address string) AccountId {
 }
 
+// BuildAccountID creates an AccountID with the provided address.
+func BuildAccountID(address string) (AccountId, error) {
```
I just added `AddressToAccountId()` to the release branch

abuiles replied:
@tamirms  perfect! I fixed it to use that, also I had an unit test for it, so I rename it to test this function.

tamirms commented on xdr/asset.go:
```diff
@@ -32,6 +32,43 @@ func MustNewCreditAsset(code string, issuer string) Asset {
+//		- `credit_alphanum4`
+//		- `credit_alphanum12`
+func BuildAsset(assetType, issuer, code string) (Asset, error) {
+	t, ok := map[string]AssetType{
```
I added `AssetTypeToString` to the release branch in this file. You can declare this map next to that definition

abuiles replied:
@tamirms good idea! Added

tamirms commented on services/horizon/internal/actions/offer.go:
```diff
@@ -55,6 +57,7 @@ func (handler GetOffersHandler) GetResourcePage(r *http.Request) ([]hal.Pageable
 	return offers, nil
 }
 
+//
```
this line can be removed

abuiles commented:
@tamirms @bartekn Addressed  all your  comments, as for the validations on assets,  I added checks for the following matrix, where only the first  one is valid (for issued assets):

|Type|Code|Issuer|
|----|----|------|
|X   | X  |  X   |
|X   | X  |      |
|X   |    |      |
|    | X  |      |
|    | X  |  X   |
|    |    |  X   |

abuiles commented on services/horizon/internal/actions/helpers.go:
```diff
@@ -711,3 +714,125 @@ func FullURL(ctx context.Context) *url.URL {
+		}
+	}
+
+	decoder.IgnoreUnknownKeys(true)
```
since decoder is  a package global, I opted to call this here, so we don't have to worry if other consumer set it back to false for their specific scenario.
